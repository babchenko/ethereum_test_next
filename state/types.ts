import { Citizen } from 'types';

export type InitialState = {
    pageData: { [key: string]: Citizen[] },
    total: number,
    loading: boolean,
    notes: { [key: number]: string },
}

export enum ActionType {
    SET_CITIZEN_PAGE = 'SET_CITIZEN_PAGE',
    IS_LOADING = 'IS_LOADING',
    SET_TOTAL = 'SET_TOTAL',
    SET_NOTE = 'SET_NOTE',
    ADD_CITIZEN = 'ADD_CITIZEN',
};


export type Action =
    | { type: ActionType.SET_CITIZEN_PAGE, payload: { page: number, items: Citizen[] } }
    | { type: ActionType.ADD_CITIZEN, payload: Citizen }
    | { type: ActionType.IS_LOADING, payload: boolean }
    | { type: ActionType.SET_TOTAL, payload: number }
    | { type: ActionType.SET_NOTE, payload: { id: number, note: string } }

