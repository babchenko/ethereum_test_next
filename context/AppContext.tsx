import React, { createContext, Dispatch, useReducer } from 'react';
import { initialState } from 'state/initialState';
import { Action, InitialState } from 'state/types';
import { reducer } from 'state/reducer';

export const AppContext = createContext<{
    state: InitialState;
    dispatch: Dispatch<Action>;
}>({
    state: initialState,
    dispatch: () => null
});

type AppProviderProps = {
    children: React.ReactNode
}

const AppProvider: React.FC<AppProviderProps> = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, initialState);

    return (
        <AppContext.Provider value={{state, dispatch}}>
            {children}
        </AppContext.Provider>
    );
}

export default AppProvider;