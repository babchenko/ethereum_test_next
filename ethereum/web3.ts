// @ts-nocheck
import Web3 from 'web3';
let web3: Web3;

// @ts-ignore
if (typeof window !== "undefined" && typeof window.ethereum !== "undefined") {
    // We are in the browser and metamask is running.
    // @ts-ignore
    window.ethereum.request({ method: "eth_requestAccounts" });
    // @ts-ignore
    web3 = new Web3(window.ethereum);
} else {
    // We are on the server *OR* the user is not running metamask
    const provider = new Web3.providers.WebsocketProvider(
        process.env.NEXT_PUBLIC_INFURA_WSS_ROPSTEN
    );

    web3 = new Web3(provider);
}

export default web3;