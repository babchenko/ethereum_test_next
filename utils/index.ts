import { EventData } from 'web3-eth-contract';
import { Citizen } from 'types';

export const getCitizen = ({ returnValues }: EventData) => {
    const citizen: Citizen = {
        id: returnValues.id,
        age: returnValues.age,
        city: returnValues.city,
        name: returnValues.name
    }

    return citizen;
}