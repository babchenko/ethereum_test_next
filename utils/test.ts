export const createClickEvent = (): MouseEvent => new MouseEvent('click', { bubbles: true });
export const sleep = (time: number) => new Promise(resolve => setTimeout(resolve, time));
export const getTracer = (search: string): string => `[data-testid="${search}"]`;