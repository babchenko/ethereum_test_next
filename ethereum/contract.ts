// @ts-nocheck
import web3 from './web3';
import abi from './abi.json';

const contractInstance = new web3.eth.Contract(
    abi,
    process.env.NEXT_PUBLIC_CONTRACT_ID
);

export default contractInstance;