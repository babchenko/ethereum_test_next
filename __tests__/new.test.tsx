import New from 'pages/new';
import { createMockRouter } from './mock/createRouter';
import { RouterContext } from 'next/dist/shared/lib/router-context';
import { AppContext } from 'context/AppContext';
import { NewPageTestProps } from '../types';
import { initialStateMock } from './mock/mockData';

import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { mount, configure } from 'enzyme';
import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';
import * as ReackHookForm from 'react-hook-form';
import web3 from '../ethereum/web3';
import contract from '../ethereum/contract';
import { createClickEvent, sleep } from '../utils/test';

configure({adapter: new Adapter()});

jest.mock("next/link", () => {
    const Link: React.FC<{ children: React.ReactElement }> = ({children}) => {
        return children;
    }
    return Link;
});

describe('New', () => {
    const dispatch = jest.fn();

    const TEST_NAME = 'fake-name';
    const TEST_AGE = "20";
    const TEST_CITY = 'fake-city';
    const TEST_NOTE = 'fake-note';
    const TEST_ETH_ACCOUNT = 'fake-eth-account';

    const Container: React.FC<NewPageTestProps> = ({initialState}) => (
        <AppContext.Provider value={ {state: initialState, dispatch} }>
            <RouterContext.Provider value={ createMockRouter({}) }>
                <New />
            </RouterContext.Provider>
        </AppContext.Provider>
    );

    it('Should match snapshot', () => {
        const { container } = render(<Container initialState={initialStateMock} />);
        expect(container).toMatchSnapshot();
    });

    it('Should display errors when form not valid', () => {
        jest.spyOn(ReackHookForm, 'useForm').mockReturnValueOnce({
            register: jest.fn(),
            handleSubmit: jest.fn(),
            // @ts-ignore
            formState: {
                errors: {
                    name: {
                        type: 'required'
                    },
                    age: {
                        type: 'required'
                    }
                }
            },
            reset: jest.fn()
        });

        const wrapper = mount(<Container initialState={initialStateMock} />);

        wrapper.find(`button[type="submit"]`).simulate('click');

        const errors = wrapper.find('.error.message');

        expect(errors.length).toBe(2);
    });

    it ('Fill form and check call to web3', async () => {

        await act(async () => {
            jest.spyOn(web3.eth, 'getAccounts').mockReturnValueOnce(Promise.resolve([TEST_ETH_ACCOUNT]))
            jest.spyOn(contract.methods, 'addCitizen').mockImplementationOnce(jest.fn());

            const { getByTestId } = render(<Container initialState={initialStateMock} />);

            const name = getByTestId('name');
            const age = getByTestId('age');
            const city = getByTestId('city');
            const note = getByTestId('note');
            const button = getByTestId('submit')

            fireEvent.change(name, { target: { value: TEST_NAME } })
            fireEvent.change(age, { target: { value: TEST_AGE } })
            fireEvent.change(city, { target: { value: TEST_CITY } })
            fireEvent.change(note, { target: { value: TEST_NOTE } })


            fireEvent(button, createClickEvent());

            await sleep(100);

            expect(contract.methods.addCitizen).toBeCalledWith(TEST_AGE, TEST_CITY, TEST_NAME, TEST_NOTE);
        })


    })
})
