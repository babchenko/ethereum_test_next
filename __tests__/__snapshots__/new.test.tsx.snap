// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`New Should match snapshot 1`] = `
<div>
  <div>
    <div
      class="ui container"
    >
      <div
        class="ui menu"
        style="margin-top: 10px;"
      >
        <a
          class="item"
        >
          FETestTask
        </a>
        <div
          class="right menu"
        >
          <a
            class="item"
          >
            +
          </a>
        </div>
      </div>
      <h3>
        Create a Citizen
      </h3>
      <form
        class="ui form"
      >
        
        
        <div
          class="field"
        >
          <label>
            Name
          </label>
          <input
            data-testid="name"
            name="name"
            placeholder="Name"
          />
        </div>
        <div
          class="field"
        >
          <label>
            City
          </label>
          <input
            data-testid="city"
            name="city"
            placeholder="City"
          />
        </div>
        <div
          class="field"
        >
          <label>
            Age
          </label>
          <input
            data-testid="age"
            max="150"
            min="18"
            name="age"
            placeholder="Age"
            step="1"
            type="number"
          />
        </div>
        <div
          class="field"
        >
          <label>
            Note
          </label>
          <input
            data-testid="note"
            name="note"
            placeholder="Note"
            type="text"
          />
        </div>
        <button
          class="ui primary button"
          data-testid="submit"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>
</div>
`;
